base_layer/core/src/chain_storage/memory_db/memory_db.rs

```diff
@@ -713,3 +731,18 @@ fn rewind_checkpoint_index(cp_count: usize, steps_back: usize) -> usize {
         1
     }
 }
+
+// Calculate the total leaf node count upto a specified height.
+fn count_mmr_nodes_added(checkpoints: &MemDbVec<MerkleCheckPoint>, height: u64) -> Result<u32, ChainStorageError> {
+    let mut node_count: u32 = 0;
+    let last_index = min(checkpoints.len()?, (height + 1) as usize);
+    for cp_index in 0..last_index {
+        if let Some(cp) = checkpoints
+            .get(cp_index)
+            .map_err(|e| ChainStorageError::AccessError(format!("Checkpoint error: {}", e.to_string())))?
+        {
+            node_count += cp.nodes_added().len() as u32;
+        }
+    }
```

Comment:
Remote nodes require the ability to determine the total number of added leaf nodes at a specific chain height, adding of UTXOs requires the same ability but at the tip of the chain. Currently, this calculation is expensive as all the checkpoints need to be traversed. A possible solution is to store the total accumulated added node count with each checkpoint to allow efficient retrieval. It could be implemented as a separate db for each checkpoint set, but then special care needs to be taken to ensure it is kept consistent during reorgs and rewinds. Another possibility is to make use of the existing checkpoint dbs, this will ensure no reorg and rewind logic need to be added but will break backwards compatibility.